Inbox

Sjmarf opened the pull request:
Implements the inbox and the tab bar unread count badge. Only a single page of the inbox is loaded at the moment.

The inbox count is refreshed every 30 seconds (closes #855). If the unread count is refreshed and the count changes, the user is prompted to refresh the inbox:

<img src="https://github.com/mlemgroup/mlem/assets/78750526/75f358b6-cb14-4e15-a86a-37015dddea09" width=50%>
<br/><br/>
I also tweaked the `menuActions` definitions to use a new `ActionBuilder` result builder to make the syntax a little nicer.

Sjmarf committed:
Update

Sjmarf committed:
Update MlemApp.swift

Sjmarf committed:
Update

Sjmarf committed:
Update

Sjmarf committed:
Update

Sjmarf committed:
Palette fixes

Sjmarf committed:
More palette fixes

Sjmarf committed:
Merge branch 'dev2' into sjmarf/inbox

Sjmarf committed:
fix

Sjmarf committed:
Merge branch 'dev2' into sjmarf/inbox

Sjmarf committed:
update

Sjmarf committed:
Update CustomTabItem.swift

Sjmarf committed:
Add ellipses menus

Sjmarf committed:
MlemMiddleware version

Sjmarf committed:
Back to branch middleware

EricBAndrews commented on Mlem.xcodeproj/project.pbxproj:
```diff
@@ -1901,8 +1937,8 @@
-				kind = upToNextMinorVersion;
-				minimumVersion = 0.14.0;
+				branch = sjmarf/inbox;
+				kind = branch;
```
Reminder

EricBAndrews commented on Mlem/App/Constants/Icons.swift:
```diff
@@ -44,6 +44,10 @@ enum Icons {
     
+    // inbox
+    static let mention: String = "quote.bubble"
+    static let message = "envelope"
```
*nit* Consistency

```suggestion
    static let message: String = "envelope"
```

EricBAndrews commented on Mlem/App/Views/Root/MlemApp.swift:
```diff
@@ -5,14 +5,35 @@
+        imageConfig.imageDecompressingQueue = OperationQueue(maxConcurrentCount: 8)
+        
+        // I'm leaving that here for mormaer, once I get a handle on rate limites that's where we put em!
+        // imageConfig.isRateLimiterEnabled
```
```suggestion
        // TODO: rate limiting
```

EricBAndrews commented on Mlem/App/Views/Root/Tabs/Inbox/InboxView.swift:
```diff
@@ -10,46 +10,221 @@ import MlemMiddleware
+                        }
+                    }
+                } header: {
+                    BubblePicker(
+                        Tab.allCases,
+                        selected: $selectedTab,
+                        label: { tab in
+                            tab.rawValue.capitalized
+                        },
+                        value: { tab in
+                            if let unreadCount = (appState.firstSession as? UserSession)?.unreadCount {
+                                switch tab {
+                                case .all:
+                                    return unreadCount.total
+                                case .replies:
+                                    return unreadCount.replies
+                                case .mentions:
+                                    return unreadCount.mentions
+                                case .messages:
+                                    return unreadCount.messages
+                                }
+                            }
+                            return 0
+                        }
+                    )
+                    .background(palette.background.opacity(isAtTop ? 1 : 0))
+                    .background(.bar)
```
*nit*

This is a pretty hefty view, worth moving to an `@ViewBuilder var header` helper

EricBAndrews commented on Mlem/App/Views/Shared/ExpandedPostView.swift:
```diff
@@ -14,37 +14,34 @@ struct ExpandedPostView: View {
-                        } catch {
-                            handleError(error)
-                        }
+                let showLoadingSymbol = showCommentWithId == nil || self.post.isUpgraded && loadingState != .loading
```
*nit*

Some parentheses are nice for clarity here, since `&&` has precedence over `||` but appears later

```suggestion
                let showLoadingSymbol = showCommentWithId == nil || (self.post.isUpgraded && loadingState != .loading)
```

EricBAndrews commented on Mlem/App/Views/Shared/ExpandedPostView.swift:
```diff
@@ -54,16 +51,55 @@ struct ExpandedPostView: View {
+            }
+            .onAppear {
+                if let showCommentWithId {
+                    // The scroll destination isn' always accurate. Possibly due to the post image changing
+                    // size on load? Using `anchor: .top` would be better here, but `anchor: .center` makes
+                    // the inaccuracy less noticeable. See also the comment further up the file.
+                    proxy.scrollTo(showCommentWithId, anchor: .center)
+                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
+                        self.showCommentWithId = nil
+                    }
```
This solution seems good for now--when we have the image size from `PostView`, we should be able to set the image size correctly before it loads and thereby fix this properly

EricBAndrews reviewed (changes requested):
Code looks good! A couple very minor nits.

I've noticed a few rough edges:

- This behavior, which should be fixed with the addition of a `FeedHeaderView`
  <img width="462" alt="Screenshot 2024-07-13 at 7 23 12 PM" src="https://github.com/user-attachments/assets/12427337-342e-44d2-ac18-545cff3e921a">
- Messages from yourself are always marked as unread and can be marked read--they should be default read and the toggle disabled
- Inbox items are not automatically marked read on interaction

I also think the blue highlight could fade a little quicker? It's pretty sluggish, though there's a fine line there

Sjmarf committed:
Apply suggestions from code review

Co-authored-by: Eric Andrews <eric.b.andrews.auto@protonmail.com>

Sjmarf committed:
Implement requested changes

Sjmarf commented:
> This behavior, which should be fixed with the addition of a FeedHeaderView

Added `FeedHeaderView` 👍 

> Messages from yourself are always marked as unread and can be marked read--they should be default read and the toggle disabled

Fixed

> Inbox items are not automatically marked read on interaction

Sure, I didn't realise v1 had this. Should we mark as read when the user taps on an inbox item to open the post?

> I also think the blue highlight could fade a little quicker? It's pretty sluggish, though there's a fine line there

Sure, I've changed the fade-out time from 1.5 to 0.4 seconds. Is that better?

Sjmarf committed:
Merge branch 'dev2' into sjmarf/inbox

Sjmarf committed:
MlemMiddleware version

Sjmarf committed:
Update

Sjmarf commented:
Added the inbox toolbar changes 👍

EricBAndrews commented on Mlem/App/Views/Root/Tabs/Inbox/InboxView.swift:
```diff
@@ -108,8 +95,31 @@ struct InboxView: View {
+                    feedDescription: .init(
+                        label: "Inbox",
+                        subtitle: "Replies, mentions and messages",
+                        color: \.inbox,
```
This line causes build failures for me on Xcode 15

```suggestion
                        color: { $0.inbox },
```

EricBAndrews reviewed (changes requested):
Looks good! The faster fade is much better.

One minor build issue.

Re: mark read on navigate, it could go either way--I think keeping it as-is (not marking on navigate) is fine, since it's not an explicit interaction.

Sjmarf committed:
Update Mlem/App/Views/Root/Tabs/Inbox/InboxView.swift

Co-authored-by: Eric Andrews <eric.b.andrews.auto@protonmail.com>

EricBAndrews reviewed (approved):
LGTM! 🚀
